src/solc_0.8/defi/rules/ContributionRules.sol

```diff
@@ -0,0 +1,194 @@
+    }
+
+    function setLandMultiplierContract(address newLandContract) external onlyOwner {
+        require(newLandContract.isContract(), "ContributionRules: invalid contract address");
+        landMultiplierContract = IERC721(newLandContract);
+    }
+
+    function addERC1155MultiplierList(
+        address contractERC1155,
+        uint256[] memory ids,
+        uint256[] memory multiplier
+    ) external onlyOwner {
+        require(contractERC1155.isContract(), "ContributionRules: invalid contract address");
+        require(ids.length > 0, "ContributionRules: invalid ids array");
+        require(multiplier.length > 0, "ContributionRules: invalid multiplier array");
+
+        IERC1155 multContract = IERC1155(contractERC1155);
+
+        _listERC1155[multContract].ids = ids;
+        _listERC1155[multContract].multiplier = multiplier;
+
+        if (_islistERC1155Member(multContract) == false) {
+            _listERC1155Index.push(multContract);
+            _listERC1155[multContract].index = _listERC1155Index.length - 1;
+        }
+    }
+
+    function addERC721MultiplierList(
+        address contractERC721,
+        bool balanceOf,
+        uint256[] memory ids
+    ) external onlyOwner {
+        require(contractERC721.isContract(), "ContributionRules: invalid contract address");
+        require(
+            (balanceOf == true && ids.length == 0) || (balanceOf == false && ids.length > 0),
+            "ContributionRules: invalid ids array"
+        );
+
+        IERC721 multContract = IERC721(contractERC721);
+
+        if (balanceOf == false) {
+            _listERC721[multContract].ids = ids;
+        }
+
+        if (_islistERC721Member(multContract) == false) {
+            _listERC721Index.push(multContract);
+            _listERC721[multContract].index = _listERC1155Index.length - 1;
+        }
+    }
+
+    function getMultiplierERC721List(address reqContract) external view returns (MultiplierERC721 memory) {
+        require(reqContract.isContract(), "ContributionRules: invalid contract address");
+        return _listERC721[IERC721(reqContract)];
+    }
+
+    function getMultiplierERC1155List(address reqContract) external view returns (MultiplierERC1155 memory) {
+        require(reqContract.isContract(), "ContributionRules: invalid contract address");
+        return _listERC1155[IERC1155(reqContract)];
+    }
+
+    function delMultiplierERC721fromList(IERC721 reqContract) external onlyOwner {
+        require(reqContract != IERC721(address(0)), "ContributionRules: invalid contract address");
+        require(_islistERC721Member(reqContract), "ContributionRules: contract is not in the list");
+        uint256 indexToDelete = _listERC721[reqContract].index;
+        IERC721 addrToMove = _listERC721Index[_listERC721Index.length - 1];
+        _listERC721Index[indexToDelete] = addrToMove;
+        _listERC721[addrToMove].index = indexToDelete;
+        _listERC721Index.pop();
+    }
```

Comment:
OZ lib could help you with that
https://docs.openzeppelin.com/contracts/3.x/api/utils#EnumerableMap

Reply:
the problem of using the EnumerableMap is that it requires an index. So, we gonna have to use something to handle the index in the anyway.

Reply:
Yea, but it also require an index. So, in the end, it's kind of the same, as I'll have to control an index to use as a key for the EnumerableMap.